### PR TITLE
add `case` and `case_dict` which support async as well

### PR DESCRIPTION
**Pull request by uriva**

### CODE REVIEW

File: gamla/functional_generic.py [hunk 1]
```diff
@@ -208,3 +205,45 @@ def composition_over_binary_curried(*args):
         pair_with,
     )
 )
+
+
+_first_truthy_index = compose_left(
+    enumerate,
+    curried.filter(toolz.second),
```
Comment (rachel-benh) [inline gamla/functional_generic.py:212]: Is it ok to assume 0 is false here (as a value)?

File: gamla/functional_generic.py [hunk 2]
```diff
@@ -208,3 +205,45 @@ def composition_over_binary_curried(*args):
...
+                    toolz.complement(operator.eq(None)), NoConditionMatched
+                ),
+                mappers.__getitem__,
+            )
+        ),
+        functional.star(functional.apply),
+    )
+
+
+def case(predicates_and_mappers: Tuple[Tuple[Callable, ...], Tuple[Callable, ...]]):
```
Comment (rachel-benh) [inline gamla/functional_generic.py:243]: shouldn't the typing be Tuple[Tuple[Callable, Callable], ...]?

Comment (uriva) [reply gamla/functional_generic.py:243]: No because it is exactly two elements.

File: gamla/functional_generic.py (hunk 1 above)
Comment (uriva) [reply gamla/functional_generic.py:212]: That's the usual signature of `filter`, anything truthy/falsy. The user can constraint this further by composing their filter function with a stricter condition.

**Pull request merged**